submariner/multiple-cable-drivers.md

```diff
@@ -0,0 +1,415 @@
+# Support fine-grained cluster connectivity and multiple cable drivers
+
+## Summary
+
+Submariner currently assumes full mesh connectivity by default and provisions VPN connections between all participating
+clusters. In addition, Submariner currently offers one of three cable drivers: VXLAN, IPSec or wireguard with limited
+configuration options. Only one of these cable drivers is enabled across all the clusters in a cluster set, that is all
+clusters that join a broker must use the same cable driver. The goal of this proposal is to outline an extended model
+that will support:
+
+- The selection of which clusters to connect and in what fashion (full mesh or connection to subset of clusters in
+  a ClusterSet).
+- The selection of one or more cable drivers in order to support different cable types between connected clusters.
+  For example, IPSec between public and private clouds but raw VXLAN between the clusters in the same cloud.
+
+There are use cases that involve supporting a mix of cable drivers or cable driver configurations within the same
+cluster set. For example, a cluster set could contain three clusters, two that are on-premise and the other
+in the cloud. Connections to the cloud cluster will need to use encryption while the on-premise clusters can use
+unencrypted connections, as they’re connected by a private link, to avoid the performance overhead of encryption.
+This could be achieved using either:
+
+- Different cable driver implementations or
+- The same cable driver implementation with differing configurations.
+
+For other use cases, it makes sense to offer finer-grained connectivity options. One such use case is a client-server
+scenario where multiple client clusters need to access a server cluster but the user does not want the client clusters
+to connect to one another. Finer-grained connectivity could also improve overall scalability where full mesh is not
+required.
+
+## Proposal
+
+This enhancement proposes a mechanism to allow users to specify their intent as to which clusters to connect and how to
+connect them. The default would still be a full mesh with a homogeneous cable driver.
+
+The standard mechanism in Kubernetes to specify intent is by creating a Kubernetes API resource whose effect is eventually
+realized into the cluster's desired state. To extend the Kubernetes API, one could either use the generic Kubernetes
+`ConfigMap` or use a custom resource definition (CRD). There's pros and cons to each but a CRD seems more suitable for
+this enhancement.
+
+This enhancement will propose a single API for finer-grained connectivity and multiple cable driver support.
+
+### Finer-grained connectivity and multiple cable driver API
+
+This following CRD is proposed:
+
+```Go
+type ClusterConnectionPolicies struct {
```

Comment:
```suggestion
type ClusterConnectionPolicy struct {
```

Shouldn't it be in the singular?

Reply:
The intention was that this holds a collection of ClusterConnectionPolicySpec - but I will update the name to a singular and make sure that the ClusterConnectionPolicySpec is a slice of ClusterConnectionPolicySpecs...